From 680ce13cacd5c376be01378b18bceecf1ca59ee7 Mon Sep 17 00:00:00 2001
From: Mario Limonciello <mario.limonciello@amd.com>
Date: Tue, 28 Nov 2023 16:52:35 -0600
Subject: [RFC 2/2] PCI: Only runtime resume devices during kexec or with
 drivers

Waking up devices as part of the PCI driver shutdown() callback is
not necessary if they don't have drivers associated or a kexec is
running.  Skip this step so that anything in a low power state can
stay there for S5.

Signed-off-by: Mario Limonciello <mario.limonciello@amd.com>
---
 drivers/pci/pci-driver.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/pci/pci-driver.c b/drivers/pci/pci-driver.c
index 57ddcc59af30..f348d06f97d7 100644
--- a/drivers/pci/pci-driver.c
+++ b/drivers/pci/pci-driver.c
@@ -507,6 +507,9 @@ static void pci_device_shutdown(struct device *dev)
 	struct pci_dev *pci_dev = to_pci_dev(dev);
 	struct pci_driver *drv = pci_dev->driver;
 
+	if (kexec_in_progress || !drv || !drv->shutdown)
+		return;
+
 	pm_runtime_resume(dev);
 
 	if (drv && drv->shutdown)
-- 
2.34.1

